### PR TITLE
Add early validation logic to dynamic_dim

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102982

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

### CODE REVIEW

**Review by angelayi — APPROVED** [review]

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -67,6 +67,19 @@
...
+    if not isinstance(t, torch.Tensor):
+        raise UserError(
+            UserErrorType.INVALID_INPUT,
+            f"Expected tensor as input to dynamic_dim but got {type(t)}"
+        )
+
+    if index >= t.dim():
+        raise UserError(
+            UserErrorType.INVALID_INPUT,
+            f"The index {index} to make dynamic is out of bounds"
```
Comment (avikchaudhuri) [inline torch/_export/__init__.py:79]: This sentence reads funny. Try:
```
Expected the dimension passed to dynamic_dim to be in the range [0...{t.dim()-1}] but got {index}, which is out of bounds for the given tensor.
```

File: torch/_export/__init__.py [hunk 2]
```diff
@@ -67,6 +67,25 @@
 #     ]
 # )
 def dynamic_dim(t: torch.Tensor, index: int):
+    if not isinstance(t, torch.Tensor):
+        raise UserError(
+            UserErrorType.INVALID_INPUT,
```
Comment (avikchaudhuri) [inline torch/_export/__init__.py:72]: Maybe it's time to document these error types. Invalid inputs can arise anywhere, e.g., I'm pretty sure some of the dynamic control flow errors are invalid input errors.

Until we rationalize these, perhaps you should rename this error type to something specific? E.g., DYNAMIC_DIM? Or update other tests as well.

**Review by avikchaudhuri — APPROVED** [review]

File: torch/_export/__init__.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/_export/__init__.py:72]: Yeah dynamic_dim makes sense. i will update it :)

**Pull request closed without merging**